source/vir/src/sst_to_air.rs

```diff
@@ -312,26 +312,83 @@ pub(crate) fn constant_to_expr(_ctx: &Ctx, constant: &crate::ast::Constant) -> E
 }
 
 #[derive(Debug, PartialEq, Eq, Clone, Copy)]
-pub(crate) enum ExprCtxt {
+pub(crate) enum ExprMode {
     Spec,
     Body,
     BodyPre,
 }
 
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub(crate) struct ExprCtxt {
+    pub mode: ExprMode,
+    pub is_bit_vector: bool,
+}
+
+fn clip_bitwise_result(bit_expr: ExprX, exp: &Exp) -> Expr {
+    if let TypX::Int(range) = &*exp.typ {
+        match range {
+            IntRange::I(_) | IntRange::ISize => {
+                return apply_range_fun(&crate::def::I_CLIP, &range, vec![Arc::new(bit_expr)]);
+            }
+            IntRange::U(_) | IntRange::USize => {
+                return apply_range_fun(&crate::def::U_CLIP, &range, vec![Arc::new(bit_expr)]);
+            }
+            _ => return Arc::new(bit_expr),
+        };
+    } else {
+        panic!("In translating Bitwise operator, encountered non-integer operand")
+    }
+}
+
+fn assert_bitvec_exp_supported(exp: &Exp) {
+    match &exp.x {
+        ExpX::Const(crate::ast::Constant::Nat(_)) => {}
+        ExpX::Var(_) => {}
+        ExpX::Binary(_, e1, e2) => {
+            assert_bitvec_exp_supported(e1);
+            assert_bitvec_exp_supported(e2);
+        }
+        ExpX::Unary(_, exp) => assert_bitvec_exp_supported(exp),
+        ExpX::UnaryOpr(UnaryOpr::Box(_) | UnaryOpr::Unbox(_), _) => {}
+        ExpX::If(e1, e2, e3) => {
+            assert_bitvec_exp_supported(e1);
+            assert_bitvec_exp_supported(e2);
+            assert_bitvec_exp_supported(e3);
+        }
+        ExpX::Bind(bnd, exp) => {
+            assert_bitvec_exp_supported(exp);
+            match &bnd.x {
+                BndX::Quant(..) | BndX::Let(_) => {}
+                _ => panic!("unsupported for bit-vector: bind conversion, {:?} ", exp.x),
+            };
+        }
+        _ => panic!("unsupported for bit-vector: expression conversion {:?}", exp.x),
+    }
+}
+
 pub(crate) fn exp_to_expr(ctx: &Ctx, exp: &Exp, expr_ctxt: ExprCtxt) -> Expr {
+    if expr_ctxt.is_bit_vector {
+        assert_bitvec_exp_supported(exp)
```

Comment:
`exp_to_expr` is already recursing `exp`, and the call to `assert_bitvec_exp_supported` introduces a nested recursion over `exp`, which is asymptotically inefficient.  I think this only needs one recursion.

Reply:
I see the new commit with `assert_bitvec_exp_supported` lifted out of the recursion, which solves the performance problem.  But does `assert_bitvec_exp_supported` need to be recursive at all?  Since `exp_to_expr` is already recursing over the whole syntax tree, can we just do a nonrecursive check inside `exp_to_expr` instead having a separate recursive traversal?

Reply:
Yes, of course. I will merge these two. 
(I was thinking I might lift `assert_bitvec_exp_supported` to incorporate a function that checks if an expression can flow into Singular, but on second thought, merging these two makes more sense. - traversing only once, not having to call the checking function)